py-rse/style.Rmd

```diff
@@ -287,755 +349,875 @@ n = count(lines)
+    a /= len(f)
+    if not values:
+        return
+    else:
+        for (i, v) in enumerate(values):
+            values[i] = v / a
 ```
 
-```text
-[-0.2, -0.5, 1.7]
+```python
+# ...after...
+def rescale_by_average(values, factors, weights):
+    if (not values) or (not factors) or (not weights):
+        return
+    a = 0.0
+    for (f, w) in zip(factors, weights):
+        a += f * w
+    a /= len(f)
+    for (i, v) in enumerate(values):
+        values[i] = v / a
 ```
 
-An equivalent special form exists for named arguments:
-the catch-all variable is conventionally called `kwargs`
-(for "keyword arguments")
-and its name is prefixed with `**` (i.e., two asterisks instead of one).
-When this is used,
-the function is given a [dictionary][dictionary] of names and values
-rather than a list:
+A related refactoring pattern is called "default and override".
+To use it,
+find cases where a value is set conditionally;
+assign the default or most common value unconditionally,
+and then override it in a special case.
+The result is fewer lines of code and clearer control flow;
+however,
+it does mean executing two assignments instead of one,
+so it shouldn't be used if the common case is expensive
+(e.g., involves a database lookup or a web request).
 
 ```python
-def set_options(tag, **kwargs):
-    result = '<{}'.format(tag)
-    for key in kwargs:
-        result += ' {}="{}"'.format(key, kwargs[key])
-    result += '/>'
-    return result
-
-print(set_options('h1', color='blue'))
-print(set_options('p', align='center', size='150%'))
+# ...before..
+if configuration['threshold'] > UPPER_BOUND:
+    scale = 0.8
+else:
+    scale = 1.0
 ```
 
-```text
-<h1 color="blue"/>
-<p align="center" size="150%"/>
+```python
+# ...after...
+scale = 1.0
+if configuration['threshold'] > UPPER_BOUND:
+    scale = 0.8
 ```
 
-Notice that the names of parameters are not quoted:
-the call is `color='blue'` and not `'color'='blue'`.
+In simple cases,
+people will sometimes put the test and assignment on a single line:
 
-## How can I pass an unknown number of parameters? {#py-rse-style-starargs}
+```python
+scale = 1.0
+if configuration['threshold'] > UPPER_BOUND: scale = 0.8
+```
 
-We can use the inverse of `*args` and `**kwargs` to match a list of values to arguments.
-In this case,
-we put the `*` in front of a list and `**` in front of a dictionary when *calling* the function,
-rather than in front of the parameter when *defining* it:
+Some programmers take this even further
+and use a [conditional expression][conditional-expression]:
 
 ```python
-def trim_value(data, low, high):
-    print(data, "with", low, "and", high)
-
-parameters = ['some matrix', 'lower bound']
-named_parameters = {'high': 'upper bound'}
-trim_value(*parameters, **named_parameters)
+scale = 0.8 if configuration['threshold'] > UPPER_BOUND else 1.0
 ```
 
-```text
-some matrix with lower bound and upper bound
-```
+However,
+this puts the default last instead of first,
```

Comment:
But you could just use `if not` to put the default first, so is this really a drawback significant enough to mention?

---

Btw, sometimes I wish Python had `unless` and `then` instead of `if not` and `else`, it would read so nicely =)

```ruby
scale = 0.8 unless configuration['threshold'] > UPPER_BOUND then 1.0
```

See how nice it looks! (turns out ruby has this, that's why the syntax highlighting is correct above)